src/java/voldemort/store/socket/clientrequest/ClientRequestExecutorFactory.java

```diff
@@ -206,26 +165,32 @@ public ClientRequestExecutor create(SocketDestination dest) throws Exception {
                                                               socketChannel,
                                                               socketBufferSize);
             int timeoutMs = this.getTimeout();
-            BlockingClientRequest<String> clientRequest = new BlockingClientRequest<String>(new ProtocolNegotiatorClientRequest(dest.getRequestFormatType()),
-                                                                                            timeoutMs);
-            clientRequestExecutor.addClientRequest(clientRequest, timeoutMs, 0);
 
-            selectorManager.registrationQueue.add(clientRequestExecutor);
-            selector.wakeup();
+            ProtocolNegotiatorClientRequest protocolRequest = new ProtocolNegotiatorClientRequest(dest.getRequestFormatType());
 
-            // Block while we wait for protocol negotiation to complete. May
-            // throw interrupted exception
-            boolean awaitResult = clientRequest.await();
-
-            if(awaitResult == false) {
-                String errorMessage = "Protocol negotiation timed out for socket "
-                                      + socketChannel.socket();
-                throw new TimeoutException(errorMessage);
-            } else {
-                // Either returns uninteresting token, or throws exception if
-                // protocol negotiation failed.
-                clientRequest.getResult();
-            }
+            NonblockingStoreCallback callback = new NonblockingStoreCallback() {
+
+                @Override
+                public void requestComplete(Object result, long requestTime) {
+                    if(result instanceof Exception) {
+                        logger.info("Reporting exception to pool " + result.getClass());
+                        pool.reportException(dest, (Exception) result);
+                    }
+                }
+
+            };
+
+            NonblockingStoreCallbackClientRequest<String> clientRequest = new NonblockingStoreCallbackClientRequest<String>(pool,
+                                                                                                                            dest,
+                                                                                                                            protocolRequest,
+                                                                                                                            clientRequestExecutor,
+                                                                                                                            callback,
+                                                                                                                            stats);
+
+            clientRequestExecutor.setConnectRequest(clientRequest, timeoutMs);
```

Comment:
The actual socket gets registered to the same selector manager. Please look at the line below selectorManager.add(clientRequestExecutor); This registers the socket to the same Selector Manager.